core/unix/module_elf.h

```diff
@@ -154,7 +163,7 @@ app_pc
 elf_loader_map_phdrs(elf_loader_t *elf, bool fixed, map_fn_t map_func,
                      unmap_fn_t unmap_func, prot_fn_t prot_func,
                      check_bounds_fn_t check_bounds_func, memset_fn_t memset_func,
-                     modload_flags_t flags);
+                     modload_flags_t flags, remap_fn_t remap_func);
 
 /* Iterate program headers of a mapped ELF image and find the string that
  * PT_INTERP points to.  Typically this comes early in the file and is always
```

Comment:
Is it possible to add a test?  The race is hard to set up; a unit test?  Maybe hard to test the bookkeeping in a unit test which is for full DR.

Reply:
Perhaps there could be such a unit test for dr_map_executable_file. We can have parallel threads that keep doing mmaps with PROT_NONE then verify that the region is indeed unreadable and munmap them; the dr_map_executable_file bug would cause the overlapping region to be readable. But I don't know how to reliably (to any degree) cause the race, since the other threads would have to mmap the same region as the anon map made by dr_map_executable_file (with no way to get that information from the thread running dr_map_executable_file before that function returns).

FWIW the internal unit test does reproduce the issue (2% of the time).